core/notification/service.go

```diff
@@ -0,0 +1,138 @@
+package notification
+
+import (
+	"context"
+
+	"github.com/odpf/salt/log"
+	"github.com/odpf/siren/core/receiver"
+	"github.com/odpf/siren/core/subscription"
+	"github.com/odpf/siren/core/template"
+	"github.com/odpf/siren/pkg/errors"
+	"gopkg.in/yaml.v3"
+)
+
+//go:generate mockery --name=SubscriptionService -r --case underscore --with-expecter --structname SubscriptionService --filename subscription_service.go --output=./mocks
+type SubscriptionService interface {
+	MatchByLabels(ctx context.Context, labels map[string]string) ([]subscription.Subscription, error)
+}
+
+//go:generate mockery --name=ReceiverService -r --case underscore --with-expecter --structname ReceiverService --filename receiver_service.go --output=./mocks
+type ReceiverService interface {
+	Get(ctx context.Context, id uint64) (*receiver.Receiver, error)
+}
+
+// NotificationService is a service for notification domain
+type NotificationService struct {
+	logger              log.Logger
+	q                   Queuer
+	receiverService     ReceiverService
+	subscriptionService SubscriptionService
+	receiverPlugins     map[string]Notifier
+}
+
+// NewService creates a new notification service
+func NewService(
+	logger log.Logger,
+	q Queuer,
+	receiverService ReceiverService,
+	subscriptionService SubscriptionService,
+	receiverPlugins map[string]Notifier,
+) *NotificationService {
+	return &NotificationService{
+		logger:              logger,
+		q:                   q,
+		receiverService:     receiverService,
+		subscriptionService: subscriptionService,
+		receiverPlugins:     receiverPlugins,
+	}
+}
+
+func (ns *NotificationService) getReceiverPlugin(receiverType string) (Notifier, error) {
+	receiverPlugin, exist := ns.receiverPlugins[receiverType]
+	if !exist {
+		return nil, errors.ErrInvalid.WithMsgf("unsupported receiver type: %q", receiverType)
+	}
+	return receiverPlugin, nil
+}
+
+func (ns *NotificationService) DispatchDirect(ctx context.Context, n Notification, receiverID uint64) error {
```

Comment:
```suggestion
func (ns *NotificationService) DispatchToReceiver(ctx context.Context, n Notification, receiverID uint64) error {
```

Reply:
I was also thinking of `DispatchToReceiver` but I imagine a case to send notification to multiple receivers in the future so that is why I picked `DispatchDirect`.

But I think `DispatchToReceiver` does make sense for now.